plugins/user_mapping/src/local_file.cpp

```diff
@@ -0,0 +1,122 @@
+#include "interface.h"
+#include "local_file.hpp"
+
+#include <cstdlib>
+#include <cstring>
+#include <exception>
+#include <fstream>
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+
+#include <spdlog/spdlog.h>
+
+namespace
+{
+	std::string file_path;
+	std::vector<UserProfile> profile_list;
+	std::shared_mutex list_mutex;
+} // anonymous namespace
+
+auto init(const nlohmann::json& _config) -> void {
+	const auto path{_config.find("file_path")};
+	if (path == std::end(_config)) {
+		// handle bad case
+		spdlog::error("{}: Unable to find 'file_path' in provided config.", __func__);
+	}
+
+	// Save file string
+	file_path = *path;
+
+	// Parse into json object
+	spdlog::trace("{}: openning file_path [{}].", __func__, file_path);
+	update();
+}
+
+auto update() -> void {
+	try {
+		std::unique_lock update_lock{list_mutex, std::defer_lock};
+
+		std::ifstream file{file_path};
+		spdlog::debug("{}: Opened file [{}]", __func__, file_path);
+
+		auto json_data{nlohmann::json::parse(file)};
+
+		// Create temp list to be swapped in later
+		std::vector<UserProfile> temp_list;
+		temp_list.reserve(json_data.size());
+
+		// Process into list:
+		auto base_itter{json_data.items()};
+		std::transform(std::begin(base_itter), std::end(base_itter), std::back_inserter(temp_list), [](const auto& _itter) -> UserProfile {
+			return {.irods_user_name = _itter.key(), .attributes = _itter.value()};
+		});
+
+		// Move new item list to old one
+		if (update_lock.try_lock()) {
+			profile_list = std::move(temp_list);
+		}
+	} catch(const std::exception& e) {
+		spdlog::debug("{}: {}", __func__, e.what());
+	}
+}
+
+auto match(const nlohmann::json& _params) -> std::optional<std::string> {
+	std::shared_lock read_lock{list_mutex, std::defer_lock};
+	update();
+
+	read_lock.lock();
+
+	// Use provided mappings to see if there is a complete match to a user
+	for (const auto& [irods_username, attributes] : profile_list) {
+		auto attr_iter{attributes.items()};
+
+		// Verify that each attribute specified is found in _params
+		auto res{std::all_of(std::begin(attr_iter), std::end(attr_iter), [&_params](const auto& _iter) -> bool {
+			const auto value_of_interest{_params.find(_iter.key())};
+			if (value_of_interest == std::end(_params)) {
+				return false;
+			}
+
+			return *value_of_interest == _iter.value();
+		})};
+
+		// If all specified attributes matched, _params maps to irods_username
+		if (res) {
+			return {irods_username};
+		}
+	}
+
+	return std::nullopt;
+}
+
+void user_mapper_init(const char* _args) {
+	try {
+		spdlog::debug("{}: recieved _args [{}]", __func__, _args);
+		init(nlohmann::json::parse(_args));
+	} catch (const std::exception& e) {
+		spdlog::debug("{}: {}", __func__, e.what());
+	}
+}
+
+char* user_mapper_match(const char* _param) {
```

Comment:
`match`, `update`, and `init` use the `auto func() -> return_type` form.

Reply:
Done.